numpyro/distributions/continuous.py

```diff
@@ -1,441 +1,373 @@
+    def __init__(self, df, validate_args=None):
+        self.df = df
+        super(Chi2, self).__init__(0.5 * df, 0.5, validate_args=validate_args)
 
-    def _rvs(self, b):
-        return random.pareto(self._random_state, b, shape=self._size)
 
-    def _cdf(self, x, b):
-        return 1 - x ** (-b)
+class Exponential(Distribution):
+    reparametrized_params = ['rate']
+    arg_constraints = {'rate': constraints.positive}
+    support = constraints.positive
 
-    def _ppf(self, q, b):
-        return np.pow(1 - q, -1.0 / b)
+    def __init__(self, rate, validate_args=None):
+        self.rate = rate
+        super(Exponential, self).__init__(batch_shape=np.shape(rate), validate_args=validate_args)
 
-    def _sf(self, x, b):
-        return x ** (-b)
+    def sample(self, key, size=()):
+        u = random.uniform(key, shape=size + self.batch_shape)
+        return -np.log1p(-u) / self.rate
 
-    def _stats(self, b, moments='mv'):
-        mu, mu2, g1, g2 = None, None, None, None
-        if 'm' in moments:
-            mask = b > 1
-            bt = np.extract(mask, b)
-            mu = np.where(mask, bt / (bt - 1.0), np.inf)
-        if 'v' in moments:
-            mask = b > 2
-            bt = np.extract(mask, b)
-            mu2 = np.where(mask, bt / (bt - 2.0) / (bt - 1.0) ** 2, np.inf)
-        if 's' in moments:
-            mask = b > 3
-            bt = np.extract(mask, b)
-            vals = 2 * (bt + 1.0) * np.sqrt(bt - 2.0) / ((bt - 3.0) * np.sqrt(bt))
-            g1 = np.where(mask, vals, np.nan)
-        if 'k' in moments:
-            mask = b > 4
-            bt = np.extract(mask, b)
-            vals = (6.0 * np.polyval([1.0, 1.0, -6, -2], bt)
-                    / np.polyval([1.0, -7.0, 12.0, 0.0], bt))
-            g2 = np.where(mask, vals, np.nan)
-        return mu, mu2, g1, g2
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        return np.log(self.rate) - self.rate * value
+
+    @property
+    def mean(self):
+        return np.reciprocal(self.rate)
+
+    @property
+    def variance(self):
+        return np.reciprocal(self.rate ** 2)
+
+
+class HalfCauchy(TransformedDistribution):
+    reparametrized_params = ['scale']
+    arg_constraints = {'scale': constraints.positive}
+    support = constraints.positive
 
-    def _entropy(self, c):
-        return 1 + 1.0 / c - np.log(c)
+    def __init__(self, scale, validate_args=None):
+        base_dist = Cauchy(0, scale)
+        self.scale = scale
+        super(HalfCauchy, self).__init__(base_dist, AbsTransform(),
+                                         validate_args=validate_args)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        log_prob = self.base_dist.log_prob(value) + np.log(2)
+        value, log_prob = promote_shapes(value, log_prob)
+        log_prob = np.where(value < 0, -np.inf, log_prob)
```

Comment:
Hmm..this is an interesting one. I think a reasonable convention and something that we could try to satisfy is that if the values lie outside the support, the corresponding log density evaluation should give a `NaN` or `-inf` or just throw an exception (not nice, but still acceptable). Otherwise the return values could appear to be valid but are actually silent bugs. In this case, since `base_dist` is `cauchy` it will happily return finite log density for negative values which can be problematic (more so since validation check is disabled by default). But you are right that, by that token, `Exponential` should also do this masking. I don't have a good solution for this. Let us remove this, and maybe make validation check enabled by default (will do so separately).

Reply:
You are right! This is more involved than I thought because negative value can provide a valid log_prob here. Same for Exponential. I think that it is good to keep this implementation as is and do the mask for each distribution which has similar issue. :)

Reply:
Sorry, how about going with your decision (to remove this mask) and taking care of this separately? Some distributions such as LogNormal is not easy to justify at a glance.